### PR TITLE
quodlibet: clean up, alternative fix for Python 3.12

### PR DESCRIPTION
**Pull request by SFrijters**
## Description of changes

The first three commits are a cleanup/update of the derivation.
The fourth commit is a possible alternative to https://github.com/NixOS/nixpkgs/pull/326051 as a fix for https://github.com/NixOS/nixpkgs/issues/325537, but I'm not sure why I have to disable the `test_cover_path_lastfm` test.

I'm not a power user, but I can run the program and play mp3s again.

Maybe @pbogdan @paveloom have more insight?

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/audio/quodlibet/default.nix [hunk 1]
```diff
@@ -1,56 +1,60 @@
...
+  tag ? "",
 
   # build time
-, gettext
-, gobject-introspection
-, wrapGAppsHook3
+  gettext,
+  gobject-introspection,
+  wrapGAppsHook3,
+  fetchpatch,
```
Comment (paveloom) [inline pkgs/applications/audio/quodlibet/default.nix:10]:
```suggestion
```

File: pkgs/applications/audio/quodlibet/default.nix [hunk 2]
```diff
@@ -1,56 +1,60 @@
-{ lib
-, fetchFromGitHub
-, tag ? ""
+{
+  lib,
+  fetchFromGitHub,
```
Comment (paveloom) [inline pkgs/applications/audio/quodlibet/default.nix:3]:
```suggestion
  fetchFromGitHub,
  fetchpatch,
```

File: pkgs/applications/audio/quodlibet/default.nix [hunk 3]
```diff
@@ -59,79 +63,113 @@ python3.pkgs.buildPythonApplication rec {
...
+
+  dependencies =
+    with python3.pkgs;
+    [
+      feedparser
+      gst-python
+      mutagen
+      pycairo
+      pygobject3
+      setuptools
```
Comment (paveloom) [inline pkgs/applications/audio/quodlibet/default.nix:131]:
```suggestion
```

File: pkgs/applications/audio/quodlibet/default.nix [hunk 4]
```diff
@@ -59,79 +63,113 @@ python3.pkgs.buildPythonApplication rec {
...
+      "--deselect=tests/test_browsers_iradio.py::TInternetRadio::test_click_add_station"
+      # upstream does actually not enforce source code linting
+      "--ignore=tests/quality"
+    ]
+    ++ lib.optionals (python3.pythonAtLeast "3.12") [
+      # gdist module is broken with python 3.12
+      "--ignore=tests/test_po.py"
+      "--ignore=tests/test_setup.py"
+      "--deselect=tests/plugin/test_covers.py::TCovers::test_cover_path_lastfm"
+    ]
```
Comment (paveloom) [inline pkgs/applications/audio/quodlibet/default.nix:168]:
```suggestion
```

Here's a patch that imports `setuptools` before `distutils` in `gdist/__init__.py`:

```
From b3980cf4c8766815e4c13ee1695cebfbf473a2e6 Mon Sep 17 00:00:00 2001
From: Pavel Sobolev <contact@paveloom.dev>
Date: Sun, 14 Jul 2024 11:30:11 +0300
Subject: [PATCH] Import `setuptools` before importing `distutils`.

---
 gdist/__init__.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/gdist/__init__.py b/gdist/__init__.py
index 71814af49..a54aac653 100644
--- a/gdist/__init__.py
+++ b/gdist/__init__.py
@@ -31,6 +31,7 @@ Also supports setuptools but needs to be imported after setuptools
 (which does some monkey patching)
 """

+import setuptools
 import sys

 from distutils.core import setup
--
2.45.2

```

**Review by paveloom — CHANGES_REQUESTED** [review]

**Comment by SFrijters** [comment]
Thanks, that indeed fixes the `gdist` problem. I still do get a (single) test failure though:
```
quodlibet-full> =================================== FAILURES ===================================
quodlibet-full> ________________________ TCovers.test_cover_path_lastfm ________________________
quodlibet-full> tests/plugin/test_covers.py:45: in test_cover_path_lastfm
quodlibet-full>     plugin_cls = self.plugins["lastfm-cover"].cls
quodlibet-full> E   KeyError: 'lastfm-cover'
```

**Comment by paveloom** [comment]
Result of `nixpkgs-review pr 326924` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>quodlibet</li>
    <li>quodlibet-full</li>
    <li>quodlibet-full.dist</li>
    <li>quodlibet-full.doc</li>
    <li>quodlibet-without-gst-plugins</li>
    <li>quodlibet-without-gst-plugins.dist</li>
    <li>quodlibet-without-gst-plugins.doc</li>
    <li>quodlibet-xine</li>
    <li>quodlibet-xine-full</li>
    <li>quodlibet-xine-full.dist</li>
    <li>quodlibet-xine-full.doc</li>
    <li>quodlibet-xine.dist</li>
    <li>quodlibet-xine.doc</li>
    <li>quodlibet.dist</li>
    <li>quodlibet.doc</li>
  </ul>
</details>

**Comment by SFrijters** [comment]
Result of `nixpkgs-review pr 326924` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages failed to build:</summary>
  <ul>
    <li>quodlibet</li>
    <li>quodlibet-full</li>
    <li>quodlibet-full.dist</li>
    <li>quodlibet-full.doc</li>
    <li>quodlibet-without-gst-plugins</li>
    <li>quodlibet-without-gst-plugins.dist</li>
    <li>quodlibet-without-gst-plugins.doc</li>
    <li>quodlibet-xine</li>
    <li>quodlibet-xine-full</li>
    <li>quodlibet-xine-full.dist</li>
    <li>quodlibet-xine-full.doc</li>
    <li>quodlibet-xine.dist</li>
    <li>quodlibet-xine.doc</li>
    <li>quodlibet.dist</li>
    <li>quodlibet.doc</li>
  </ul>
</details>

**Comment by paveloom** [comment]
OfBorg [was able](https://logs.ofborg.org/?key=nixos/nixpkgs.326924&attempt_id=5f268d86-a2c5-4574-aec5-9364d30c3f6e) to build it, so I would say that's something with your system...

It reaches to the `dbus-daemon` because the tests are run in a virtual X server environment in a DBus session:

https://github.com/NixOS/nixpkgs/blob/0c9abca0832a507d92eef0def4bf9634605d6724/pkgs/applications/audio/quodlibet/default.nix#L175-L183

**Comment by SFrijters** [comment]
Ok, then I think this is a better fix than the Python 3.11 pin?

**Comment by paveloom** [comment]
Yeah, I think so. I thought there would be more patches required, but it turned out to be manageable.

**Comment by cab404** [comment]
I've submitted #333086, but I guess it shouldn't interfere much with this rework.

**Comment by paveloom** [comment]
@cab404 Note that these two PRs are incompatible since you're proposing to switch to the unstable version in yours.

**Comment by cab404** [comment]
> @cab404 Note that these two PRs are incompatible since you're proposing to switch to an unstable version in yours.

iiuc, this also contains several changes to packaging, which I did not incorporate

**Comment by cab404** [comment]
Result of `nixpkgs-review pr 326924` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>quodlibet</li>
    <li>quodlibet-full</li>
    <li>quodlibet-full.dist</li>
    <li>quodlibet-full.doc</li>
    <li>quodlibet-without-gst-plugins</li>
    <li>quodlibet-without-gst-plugins.dist</li>
    <li>quodlibet-without-gst-plugins.doc</li>
    <li>quodlibet-xine</li>
    <li>quodlibet-xine-full</li>
    <li>quodlibet-xine-full.dist</li>
    <li>quodlibet-xine-full.doc</li>
    <li>quodlibet-xine.dist</li>
    <li>quodlibet-xine.doc</li>
    <li>quodlibet.dist</li>
    <li>quodlibet.doc</li>
  </ul>
</details>

**Review by cab404 — APPROVED** [review]

**Pull request merged**